File: frontend/src/lib/album-grouping.test.ts
import { describe, expect, it } from 'vitest';
import {
  hydrateAlbums,
  songAddedTime,
  songLikedTime,
  sortAlbums,
  type AlbumSummary,
  type AlbumSummaryDto,
  type ApiSong
} from './api-client';

/*
 * What is left of album grouping on this side.
 *
 * The grouping itself — folder keys, the name merge, the year election, the added-date rule — moved
 * to `GET /api/albums`, and is pinned by `AlbumProjectionTests` in the API suite. It used to live
 * here in full AND in a second full copy in the Android client, which is how one added-date rule
 * came to need fixing twice. What stays here is what depends on client state: the per-song stamps
 * the track lists sort on, the grid's own ordering, and the join that turns server cards back into
 * the song rows this app holds.
 */

// Minimal ApiSong factory — only the fields these helpers read.
function song(over: Partial<ApiSong>): ApiSong {
  return {
    id: 0,
    fileName: 'track.flac',
    artist: 'Kanye West',
    albumArtist: 'Kanye West',
    album: 'My Beautiful Dark Twisted Fantasy',
    ...over
  } as ApiSong;
}

/** Minimal album card as the API sends it. */
function card(over: Partial<AlbumSummaryDto> = {}): AlbumSummaryDto {
  const artist = over.artist ?? 'Kanye West';
  const title = over.title ?? 'MBDTF';
  return {
    key: `${artist.toLowerCase()}::${title.toLowerCase()}`,
    folderKeys: [],
    nameKey: `${artist.toLowerCase()}::${title.toLowerCase()}`,
    title,
    artist,
    year: null,
    trackCount: 0,
    durationSeconds: 0,
    byteSize: 0,
    genre: null,
    label: null,
    catalogNumber: null,
    upc: null,
    releaseDate: null,
    musicBrainzReleaseId: null,
    coverSongId: null,
    addedAtUtc: null,
    playCount: 0,
    trackIds: [],
    ...over
  };
}

describe('songAddedTime', () => {
  it('prefers the immutable acquisition stamp over pipeline timestamps', () => {
    const t = songAddedTime(
      song({
        acquiredAtUtc: '2020-01-01T00:00:00Z',
        indexedAtUtc: '2026-07-01T00:00:00Z',
        libraryBuiltAtUtc: '2026-07-02T00:00:00Z'
      })
    );

    expect(t).toBe(Date.parse('2020-01-01T00:00:00Z'));
  });

  it('falls back to the oldest churn-prone stamp for rows predating the column', () => {
    // A re-tag or re-index would have bumped only one of the two; the older one is the better guess.
    const t = songAddedTime(
      song({ indexedAtUtc: '2026-07-01T00:00:00Z', libraryBuiltAtUtc: '2022-05-05T00:00:00Z' })
    );

    expect(t).toBe(Date.parse('2022-05-05T00:00:00Z'));
  });

  it('prefers an earlier Spotify save date over the download date', () => {
    // A years-old liked song the wishlist downloader only got round to fetching today: without this
    // it would sit at the top of "recently added" next to things actually just acquired.
    const t = songAddedTime(
      song({
        acquiredAtUtc: '2026-07-26T15:44:00Z',
        spotifyAddedAtUtc: '2023-03-21T00:00:00Z'
      })
    );

    expect(t).toBe(Date.parse('2023-03-21T00:00:00Z'));
  });

  it('keeps the acquisition date when the Spotify save came later', () => {
    // Ripped in 2019, saved on Spotify in 2024 — the save date must not push it forward.
    const t = songAddedTime(
      song({
        acquiredAtUtc: '2019-01-01T00:00:00Z',
        spotifyAddedAtUtc: '2024-06-01T00:00:00Z'
      })
    );

    expect(t).toBe(Date.parse('2019-01-01T00:00:00Z'));
  });

  it('uses the Spotify save date for rows with no acquisition stamp at all', () => {
    const t = songAddedTime(song({ spotifyAddedAtUtc: '2023-03-21T00:00:00Z' }));

    expect(t).toBe(Date.parse('2023-03-21T00:00:00Z'));
  });
});

describe('songLikedTime', () => {
  it('is 0 for a song that was never liked, whatever its Spotify save date', () => {
    expect(songLikedTime(song({ spotifyAddedAtUtc: '2023-03-21T00:00:00Z' }))).toBe(0);
  });

  it('prefers the Spotify save date over the bulk auto-like sweep stamp', () => {
    // The sweep stamps every song it matches with one shared `now`; without this the whole batch
    // ties and "newest liked first" collapses into the tie-break order.
    const sweptAt = '2026-07-12T21:52:00Z';
    const a = song({ likedAtUtc: sweptAt, spotifyAddedAtUtc: '2022-09-09T00:00:00Z' });
    const b = song({ likedAtUtc: sweptAt, spotifyAddedAtUtc: '2023-08-04T00:00:00Z' });

    expect(songLikedTime(a)).toBe(Date.parse('2022-09-09T00:00:00Z'));
    expect(songLikedTime(b)).toBe(Date.parse('2023-08-04T00:00:00Z'));
    expect(songLikedTime(b)).toBeGreaterThan(songLikedTime(a));
  });

  it('keeps a heart clicked here when it predates the Spotify save', () => {
    const t = songLikedTime(
      song({ likedAtUtc: '2026-01-05T00:00:00Z', spotifyAddedAtUtc: '2026-06-01T00:00:00Z' })
    );

    expect(t).toBe(Date.parse('2026-01-05T00:00:00Z'));
  });

  it('falls back to the local like time when there is no Spotify save date', () => {
    const t = songLikedTime(song({ likedAtUtc: '2026-01-05T00:00:00Z' }));

    expect(t).toBe(Date.parse('2026-01-05T00:00:00Z'));
  });
});

describe('sortAlbums', () => {
  // Cards as the API sends them — the grid orders what it is given, it no longer builds it.
  const albums = hydrated([
    card({
      artist: 'Zappa',
      title: 'Hot Rats',
      year: 1969,
      playCount: 10,
      addedAtUtc: '2020-01-01T00:00:00Z'
    }),
    card({
      artist: 'Aphex Twin',
      title: 'Drukqs',
      year: 2001,
      playCount: 1,
      addedAtUtc: '2026-01-01T00:00:00Z'
    })
  ]);
  const titles = (key: Parameters<typeof sortAlbums>[1]) =>
    sortAlbums(albums, key).map((a) => a.title);

  it('orders by each key', () => {
    expect(titles('recent')).toEqual(['Drukqs', 'Hot Rats']);
    expect(titles('artist')).toEqual(['Drukqs', 'Hot Rats']);
    expect(titles('title')).toEqual(['Drukqs', 'Hot Rats']);
    expect(titles('year')).toEqual(['Drukqs', 'Hot Rats']);
    expect(titles('played')).toEqual(['Hot Rats', 'Drukqs']);
  });

  it('breaks ties alphabetically instead of leaving the order it was handed', () => {
    const tied = hydrated([
      card({ artist: 'B', title: 'B album' }),
      card({ artist: 'A', title: 'A album' })
    ]);

    // No play counts, no added dates — every comparator ties, so artist order decides.
    expect(sortAlbums(tied, 'played').map((a) => a.artist)).toEqual(['A', 'B']);
    expect(sortAlbums(tied, 'recent').map((a) => a.artist)).toEqual(['A', 'B']);
  });
});

describe('hydrateAlbums', () => {
  it('resolves track ids to the very song objects the store holds', () => {
    // Identity, not equality: the store mutates a row in place on a heart tap, and the album views
    // only see that because they hold the same object. A copy here would break every overlay.
    const one = song({ id: 1, title: 'Gorgeous' });
    const two = song({ id: 2, title: 'Power' });
    const byId = new Map([one, two].map((s) => [s.id, s]));

    const album = hydrateAlbums([card({ trackIds: [2, 1] })], byId)[0];

    expect(album.songs).toHaveLength(2);
    expect(album.songs[0]).toBe(two);
    expect(album.songs[1]).toBe(one);
  });

  it('drops a track the songs list does not have rather than emitting a hole', () => {
    // The two fetches can straddle a library change; the next refresh reconciles.
    const byId = new Map([[1, song({ id: 1 })]]);

    expect(hydrateAlbums([card({ trackIds: [1, 99] })], byId)[0].songs.map((s) => s.id)).toEqual([1]);
  });

  it('turns the cover track id into a cover URL, and no artwork into null', () => {
    expect(hydrateAlbums([card({ coverSongId: 7 })], new Map())[0].coverUrl).toContain('7');
    expect(hydrateAlbums([card()], new Map())[0].coverUrl).toBeNull();
  });
});

function hydrated(cards: AlbumSummaryDto[]): AlbumSummary[] {
  return hydrateAlbums(cards, new Map());
}

File: android/app/src/test/java/com/musichoarder/app/data/AlbumGroupingTest.kt
package com.musichoarder.app.data

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertSame
import org.junit.Test

/**
 * What is left of album grouping on this side.
 *
 * The grouping itself — folder keys, the name merge, the year election, the added-date rule — moved
 * to `GET /api/albums` and is pinned by `AlbumProjectionTests` in the API suite. It used to live
 * here, as a hand-written port of `buildAlbumsFromSongs`, alongside the web's own copy; that is how
 * one added-date rule came to need fixing twice and how the phone and the browser came to disagree
 * about which track names a card. What stays here is the join and the grid's own ordering.
 */
class AlbumGroupingTest {

    @Test
    fun `hydration resolves track ids to the very rows the library holds`() {
        // Identity, not equality: likes and play counts are overlays keyed on these rows, and an
        // album carrying its own copies would show a stale heart until the next refresh.
        val one = song(id = 1, title = "Gorgeous")
        val two = song(id = 2, title = "Power")

        val album = hydrateAlbums(listOf(card(trackIds = listOf(2, 1))), mapOf(1 to one, 2 to two))
            .single()

        assertEquals(listOf(2, 1), album.tracks.map { it.id })
        assertSame(two, album.tracks.first())
    }

    @Test
    fun `a track the library does not have is dropped rather than left as a hole`() {
        // The two fetches can straddle a library change; the next refresh reconciles.
        val album = hydrateAlbums(listOf(card(trackIds = listOf(1, 99))), mapOf(1 to song(id = 1)))
            .single()

        assertEquals(listOf(1), album.tracks.map { it.id })
    }

    @Test
    fun `the cover comes from the server, and falls back to a local scan`() {
        val withArt = song(id = 1, hasCoverArt = true)
        val tracks = mapOf(1 to withArt)

        assertEquals(1, hydrateAlbums(listOf(card(trackIds = listOf(1), coverSongId = 1)), tracks).single().coverTrackId)
        // An older server does not name one; the joined tracks still can.
        assertEquals(1, hydrateAlbums(listOf(card(trackIds = listOf(1))), tracks).single().coverTrackId)
        assertNull(hydrateAlbums(listOf(card()), emptyMap()).single().coverTrackId)
    }

    @Test
    fun `a card with no folders of its own is still addressable by its key`() {
        assertEquals(listOf("a::b"), hydrateAlbums(listOf(card(key = "a::b")), emptyMap()).single().folderKeys)
    }

    @Test
    fun `every order falls back to artist then title, so ties stay alphabetical`() {
        // None of these has a play count, a year, or an added date, so only the fallback can order
        // them. Without it they would come out in whatever order the server happened to send.
        val albums = listOf(
            summary(artist = "Zeta", name = "Charlie"),
            summary(artist = "Alpha", name = "Delta"),
            summary(artist = "Alpha", name = "Beta"),
        )

        // Title is excluded: it is the one key whose own comparator has something to say here.
        for (key in AlbumSortKey.entries - AlbumSortKey.Title) {
            assertEquals(
                "sorted by $key",
                listOf("Beta", "Delta", "Charlie"),
                sortAlbums(albums, key).map { it.name },
            )
        }
        assertEquals(listOf("Beta", "Charlie", "Delta"), sortAlbums(albums, AlbumSortKey.Title).map { it.name })
    }

    @Test
    fun `names are collated, not compared by codepoint`() {
        // Kotlin's natural String order would file every lowercase name after every uppercase one,
        // and "Ólafur" past "Z". The server orders the same way, using the invariant culture.
        val albums = listOf(
            summary(artist = "Zola", name = "Z"),
            summary(artist = "Ólafur", name = "O"),
            summary(artist = "aphex", name = "A"),
        )

        assertEquals(listOf("aphex", "Ólafur", "Zola"), sortAlbums(albums, AlbumSortKey.Artist).map { it.artist })
    }

    @Test
    fun `most played sorts on the play count the server summed`() {
        val albums = listOf(summary(name = "Quiet", playCount = 1), summary(name = "Loud", playCount = 9))

        assertEquals(listOf("Loud", "Quiet"), sortAlbums(albums, AlbumSortKey.Played).map { it.name })
    }

    @Test
    fun `recently added sorts on the album date the server decided`() {
        val albums = listOf(
            summary(name = "Old", addedAtUtc = "2019-01-01T00:00:00Z"),
            summary(name = "New", addedAtUtc = "2026-08-01T00:00:00Z"),
        )

        assertEquals(listOf("New", "Old"), sortAlbums(albums, AlbumSortKey.Recent).map { it.name })
    }

    private fun card(
        key: String = "an artist::an album",
        coverSongId: Int? = null,
        trackIds: List<Int> = emptyList(),
    ) = AlbumSummaryDto(
        key = key,
        folderKeys = emptyList(),
        nameKey = key,
        title = "An Album",
        artist = "An Artist",
        trackIds = trackIds,
        coverSongId = coverSongId,
    )

    /** An already-joined card, for the ordering tests — they never look at the tracks. */
    private fun summary(
        artist: String = "An Artist",
        name: String = "An Album",
        year: Int? = null,
        playCount: Int = 0,
        addedAtUtc: String? = null,
    ) = hydrateAlbums(
        listOf(
            AlbumSummaryDto(
                key = "${artist.lowercase()}::${name.lowercase()}",
                nameKey = "${artist.lowercase()}::${name.lowercase()}",
                title = name,
                artist = artist,
                year = year,
                playCount = playCount,
                addedAtUtc = addedAtUtc,
            ),
        ),
        emptyMap(),
    ).single()
}
